pkg/obfuscate: fix SQL tokenizer number parsing

knusbaum opened the pull request:
scanNumber failed to parse certain negative numbers and also failed to
set an error in the tokenizer when it returned LexError, resulting in
`nil` error messages. This commit causes scanNumber to correctly parse
those numbers and also set a more meaningful error message when it
encounters an error.

This also adds some tests specifically for the number parser.

knusbaum commented:
I'll add a regression testcase for this.

gbbr commented on pkg/obfuscate/sql_tokenizer.go:
```diff
@@ -634,7 +634,7 @@ func (tkn *SQLTokenizer) scanNumber(seenDecimalPoint bool) (TokenKind, []byte) {
 			// octal int
 			if seenDecimalDigit {
-				// tkn.setErr called in caller
+				tkn.setErr("invalid number")
```
Can we make a more specific error message? Why is it an invalid number? If we end up debugging this, we'll have an easier time knowing where the error occurred vs. having two places with the same error message.

knusbaum replied:
Done.

knusbaum committed:
pkg/obfuscate: set an error in the tokenizer when we fail to parse a number

scanNumber failed to set an error in the tokenizer when it returned
LexError, resulting in `nil` error messages. This commit causes scanNumber
to set an error message in the tokenizer as soon as it encounters an error
rather than relying on the caller to do so.

knusbaum committed:
add test back to working

ajgajg1134 commented on pkg/obfuscate/sql_test.go:
```diff
@@ -1943,6 +1942,41 @@ func TestUnicodeDigit(t *testing.T) {
+	var testCases = []struct {
+		in       string
+		expected string
+		error    string
```
Error appears unused so it can be removed

knusbaum committed:
cleanup test

knusbaum committed:
add release notes

knusbaum committed:
cleanup

ajgajg1134 reviewed (approved):
Looks good!

knusbaum committed:
add one more test case
